### ISSUE TITLE
Object graph visualization using VisualizeObjectGraph produces incomplete result for IEnumerable<> dependecies

### ISSUE THREAD

**Issue by pavel-pargachev** [simpleinjector/SimpleInjector#759]
`VisualizeObjectGraph` does not include component types that are part of `IEnumerable<>` dependency.

### How To Reproduce
Consider the following registrations:
```C#
var container = new Container();

container.Collection.Register(typeof(IValidator<>), typeof(IValidator<>).Assembly);
container.Register(typeof(IValidator<>), typeof(CompositeValidator<>));
container.Register<MyService>();

container.Verify();

var actual = container.GetRegistration(typeof(MyService)).VisualizeObjectGraph();

Console.WriteLine(actual);
```

Using types:
```C#
class MyService { public MyService(IValidator<MyModel> myModelValidator) { } }

interface IValidator<T> { }

class Composite<T> : IValidator<T> { public Composite(IEnumerable<IValidator<T>> v) { } }

class Validator1 : IValidator<MyModel> { public Validator1(Dependency d) { } }
class Validator2 : IValidator<MyModel> { public Validator2(Dependency d) { } }

class Dependency { }

class MyModel { }
```

Prints to the console
```
MyService( // Transient
    Composite<MyModel>( // Transient
        IEnumerable<IValidator<MyModel>>( // Singleton
            Dependency(), // Transient
            Dependency()))) // Transient
```

### Expected to print
```
MyService( // Transient
    Composite<MyModel>( // Transient
        IEnumerable<IValidator<MyModel>>( // Singleton
            Validator1( // Transient
                Dependency()), // Transient
            Validator2( // Transient
                Dependency())))) // Transient
```

### Additional context
* SimpleInjector 4.7.1
* .NET Core 3.0 console app

**Comment by dotnetjunkie** [comment]
Thank you for sending this high-quality repro. I can confirm that this *is* a bug. I unfortunately can't give you a workaround for this issue, so I hope this isn't a blocking issue for you.

**Comment by dotnetjunkie** [comment]
[bug-759](https://github.com/simpleinjector/SimpleInjector/tree/bug-759) branch created.

**Comment by kwlin** [comment]
I think the issue is in this code:

https://github.com/simpleinjector/SimpleInjector/blob/d106e31b306e38da91855579b37eee55c93ff079/src/SimpleInjector/Internals/InstanceProducerVisualizer.cs#L61-L64

When producer is of a ServiceType IEnumerable<> it returns all dependencies within the IEnumerable<> type, instead it should iterate through the IEnumerable<> to get the producer of an item.

@dotnetjunkie: Should I check if the ServiceType is an IEnumerable<>, then iterate through the Relationships; Per Relationship find/get the InstanceProducer? Or is there another way to get the InstanceProducers within an IEnumerable<> ServiceType?
 
@pavel-pargachev: Thanks for the great reproduction steps/unittests.

**Comment by dotnetjunkie** [comment]
@kwlin, checking for just `IEnumerable<T>` is probably not enough, because there is a range of collection types that is supported.

I'm also wondering whether this code is the correct place to fix this. Perhaps the problem lies more deeply inside the `InstanceProducer` or its `Registration` instance. This bug might possibly also have effect on other parts of the system, for instance when users analyzes the relationships of a registration themselves, they would also see this same (incorrectly) flattened graph.

**Comment by dotnetjunkie** [comment]
I've been digging a bit deeper, and I think the core of the problem lies here:

https://github.com/simpleinjector/SimpleInjector/blob/d106e31b306e38da91855579b37eee55c93ff079/src/SimpleInjector/Internals/ContainerControlledCollection.cs#L127-L132

As @kwlin's code snippet shows, `InstanceProducer` instances can return a list of `KnownRelationship` instances through their `GetRelationship()` method. `KnownRelationship` describes the relationship and is used to do verification and visualizing the object graph.

A `KnownRelationship` describes the relationship between the component and its dependencies. In the case of the `MyService` component, for instance, it contains one `KnownRelationship` and its points back to `MyService` as consumer and `IValidator<MyModel>` as dependency.

This model breaks in case of working with collections, where Simple Injector's internal `ContainerControlledCollection<T>` returns `KnownRelationship` instances where the consumer is either `Validator1` or `Validator2` and the dependency the validators' dependencies.

This is clearly wrong, because the relationship is between the collection and the validators. So the `ContainerControlledCollection<T>` is removing itself from the equation. This leads to scewed results.

Okay, so now we know what the problem is. But fixing it proves to be a bit harder than I initially thought, because:
* A `KnownRelationship` wraps an `InjectionConsumerInfo`. This gives information about the consuming end of the relationship. But an `InjectionConsumerInfo` is described using either a (constructor) `ParameterInfo` or a `PropertyInfo`. The `InjectionConsumerInfo`, therefore, not only describes the type of the consumer, but also the parameter or property in which *the dependency* is injected into. But now the question becomes: where is the dependency (either `Validator1` or `Validator2`) injected into? This is something that probably has to be made up, because in reality the are not injected (instead, they are created by the `ContainerControlledCollection<T>`, but that's an implementation detail).
* Letting the `ContainerControlledCollection<T>` (or the `ConrolledCollectionHelper.ContainerControlledCollectionRegistration`) produce 'the right' `KnownRelationship` instances, however, causes a new problem; because of the addition of the validators, the diagnostic are tripped and causes lifestyle mismatches. I haven't investigated much further on what causes these mismatches and how they should be fixed. This likely requires a change in the diagnostic analyzer, but I have to investigate this.

I'll keep you posted with my progress.

**Comment by dotnetjunkie** [comment]
Wow, that was quite a ride. I've been digging deep into this again today and am happy to announce that I was able to fix the problem. And with it, several small bugs surrounding the creation of `KnownRelationship` types concerning collection elements. That part was a bit a mess. Although there were quite a few tests concerning this part of the system, horrifically enough, those tests just echoed what the system was doing, instead of verifying what the system should have been doing. In other words, there were tests that tested the wrong behavior. Nice.

I was able to fix the problem by added an internal flag to `KnownRelationship` and by letting the `ContainerControlledCollection` produce `KnownRelationship` instances that are flagged as `UseForVerification = false`. This allows the elements of a `ContainerControlledCollection` to be used for visualization purposes, while being skipped when it comes to diagnostics. In other words, there is now a distinction between “visualization only” and “visualization and verification” registrations.

https://github.com/simpleinjector/SimpleInjector/commit/926563f239efd67f22a0c09ead94a0c9767269a1 shows the complete change.

**Comment by kwlin** [comment]
Wow what a fix; I initially thought it was a simple/quick fix, but really underestimated it ;) 

Anyway I have just a small question about the [code change](https://github.com/simpleinjector/SimpleInjector/commit/926563f239efd67f22a0c09ead94a0c9767269a1#r37662391)

**Comment by dotnetjunkie** [comment]
> initially thought it was a simple/quick fix, but really underestimated it 

Tell me about it

Thanks for your watchful eye. Much appreciated.

**Comment by dotnetjunkie** [comment]
This bug has been fixed in [v4.9.1](https://github.com/simpleinjector/SimpleInjector/releases/tag/v4.9.1).
